(fix) Add `w`rite to opening mode in object writer

maxmynter opened the pull request:
Python's `open` needs a specification of one of
create/read/write/append.

Closes #7928

## Change Description
Add `w`rite mode to binary open.

### Background

Share context and relevant information for the PR: offline discussions, considerations, design decisions etc.

### Bug Fix

If this PR is a bug fix, please let us know about:

1. Problem - 
Python needs specification of create/read/write/append.
2. Root cause - Discovered root cause after investigation
The opening mode specification was missing.
3. Solution - How the bug was fixed
  Added the opening mode specification

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/treeverse/lakeFS?pullRequest=7961) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/treeverse/lakeFS?pullRequest=7961) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/treeverse/lakeFS?pullRequest=7961) it.</sub>

maxmynter commented:
@ozkatz 

Any change requests to this?

arielshaqed reviewed (commented):
Sorry for letting this slip by me and the resultant delay in reviewing.  Thanks!

Can you figure out a way to test this in clients/python-wrapper/tests/integration/test_object.py?  If we made the error once, we might make it again.  Please don't waste too much time on it, though; if you have no quick way, just open an issue.\

Again, thanks!

arielshaqed commented:
Running Esti through a clone #7978.

maxmynter commented:
I've added a test in 62c1f63, with those i found that conditionally setting the encoding results in errors if the write mode is hardcode. Binary writes throw errors with an encoding set.

I've removed the conditional in 9eadf81. Once no longer affected by the `urllib` bug, we should consider adding both back.

arielshaqed commented on clients/python-wrapper/tests/integration/test_object.py:
```diff
@@ -458,3 +458,28 @@ def test_writer_gif(setup_repo, datafiles):
+
+    writer = ObjectWriter(stored_obj, mode=w_mode)
+
+    # Create file of size greater than _WRITER_BUFFER_SIZE (32 MB)
+    # set in SpooledTemporaryFile to force disk write
+    binary_data = b"\x00" * (32 * 1024 * 1024 + 1)
```
This is good stuff but unfortunate: we should be able to set a lower _WRITER_BUFFER_SIZE for this test, for instance, by passing ObjectWriter a max_size parameter.

Out of scope for this PR, so please add
```suggestion
    # Create file of size greater than _WRITER_BUFFER_SIZE (32 MB)
    # set in SpooledTemporaryFile to force disk write
    #
    # TODO(arielshaqed/n-o-z): set a smaller spool size.
    binary_data = b"\x00" * (32 * 1024 * 1024 + 1)
```

arielshaqed reviewed (approved):
Thanks!

Let me know if you want to add the requested TODO, or want me to do it.  With that we're good to go.

MANY THANKS for the PR and your patience - it is very much appreciated.

maxmynter commented:
Happy to help :) 

All done from my side!

arielshaqed commented:
Thanks!

Sorry for delays, will pull shortly.  @N-o-Z we can address [this comment](https://github.com/treeverse/lakeFS/pull/7961#discussion_r1674126991) right after; I will open an issue.

arielshaqed commented:
System test [passes](https://github.com/treeverse/lakeFS/actions/runs/9962736938/job/27527305218?pr=7995) :-) on my copy of this branch.

Pulling this one, thanks!
